### PR TITLE
Emotions feedback collection

### PR DESCRIPTION
**Pull request by vdragan1993**
- Added backend support for collecting of emotions feedback. 
- Added integration test.

### CODE REVIEW

File: src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tutor.Core.DomainModel.Feedback
+{
+    public class EmotionsFeedback
+    {
+        public int Id { get; set; }
+        public string Comment { get; set; }
```
Comment (Luburic) [inline src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs:12]: Can this be any string or is it limited to a handful of options?

File: src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using FluentResults;
+using System.Collections.Generic;
+
+namespace Tutor.Core.DomainModel.Feedback
+{
+    public class FeedbackService : IFeedbackService
+    {
+        private readonly IFeedbackRepository _ifeedbackRepository;
```
Comment (Luburic) [inline src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs:8]: You can remove the `i` from the field name.

File: src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+        public FeedbackService(IFeedbackRepository ifeedbackRepository)
+        {
+            _ifeedbackRepository = ifeedbackRepository;
+        }
+
+        public Result<EmotionsFeedback> SaveEmotionsFeedback(EmotionsFeedback emotionsFeedback)
+        {
+            if (emotionsFeedback == null)
```
Comment (Luburic) [inline src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs:17]: For fail fast/return early we follow the convention of one line IF-s. I.e.:

if(emotionsFeedback == null) return Result.Fail("Empty Emotions Feedback");

File: src/Tutor.Web/Controllers/Domain/DTOs/Feedback/EmotionsFeedbackDTO.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Tutor.Web.Controllers.Domain.DTOs.Feedback
+{
+    public class EmotionsFeedbackDTO
```
Comment (Luburic) [inline src/Tutor.Web/Controllers/Domain/DTOs/Feedback/EmotionsFeedbackDTO.cs:3]: Change `DTO` to `Dto`.

File: src/Tutor.Web/Controllers/Domain/FeedbackController.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Tutor.Core.DomainModel.Feedback;
+using Tutor.Web.Controllers.Domain.DTOs.Feedback;
+
+namespace Tutor.Web.Controllers.Domain
+{
+    [Route("api/feedback/")]
+    [ApiController]
+    public class FeedbackController : ControllerBase
```
Comment (Luburic) [inline src/Tutor.Web/Controllers/Domain/FeedbackController.cs:12]: Don't forget to add `[Authorize(Policy = "learnerPolicy")]`.

You can view how we pack submissions on the front end to see how the learnerId is retrieved.

**Review by Luburic — CHANGES_REQUESTED** [review]
Several minor things to improve.

File: src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs (hunk 1 above)
Comment (vdragan1993) [reply src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs:12]: We expect text that's 2 or 3 sentences long here. 
Therefore, this should be enough.

File: src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs (hunk 2 above)
Comment (Luburic) [reply src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs:8]: Yeah, this is a mistake in the KCService.

File: tests/Tutor.Web.Tests/Integration/Domain/FeedbackTests.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+namespace Tutor.Web.Tests.Integration.Domain
+{
+    [Collection("Sequential")]
+    public class FeedbackTests : BaseIntegrationTest
+    {
+        public FeedbackTests(TutorApplicationTestFactory<Startup> factory) : base(factory) {}
+
+        [Theory]
+        [MemberData(nameof(EmotionsFeedbackSubmission))]
+        public void Stores_emotions_feedback(EmotionsFeedbackDto postedFeedback, EmotionsFeedbackDto expectedFeedback)
```
Comment (Luburic) [inline tests/Tutor.Web.Tests/Integration/Domain/FeedbackTests.cs:25]: I assume that all tests in the projects are passing?

File: src/Tutor.Web/Controllers/Domain/DTOs/Feedback/EmotionsFeedbackDTO.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Tutor.Web.Controllers.Domain.DTOs.Feedback
+{
+    public class EmotionsFeedbackDto
+    {
+        public string Comment { get; set; }
```
Comment (Luburic) [inline src/Tutor.Web/Controllers/Domain/DTOs/Feedback/EmotionsFeedbackDTO.cs:5]: Still missing the learner ID. I imagine you will add it next? On the front end, this code should help https://github.com/Clean-CaDET/platform-tutor-ui-web/blob/master/src/app/modules/domain/learning-objects/question/question.service.ts#L24

Also to what do we want to tie this feedback? To a specific unit or KC?
Finally, we should include a DateTime to understand when the comment was made.

**Review by Luburic — CHANGES_REQUESTED** [review]
One thing to expand.

File: tests/Tutor.Web.Tests/Integration/Domain/FeedbackTests.cs (hunk 6 above)
Comment (vdragan1993) [reply tests/Tutor.Web.Tests/Integration/Domain/FeedbackTests.cs:25]: Added test and all tests from the `Domain` sub-folder are passing. 
Only test that's falling on my machine is `Can_transform_to_sql` from `Infrastructure` sub-folder, due to the way the test is written.

File: src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tutor.Core.DomainModel.Feedback
+{
+    public class EmotionsFeedback
+    {
+        public int Id { get; set; }
```
Comment (Luburic) [inline src/Tutor.Core/DomainModel/Feedback/EmotionsFeedback.cs:11]: Define private setters for everything by default.

File: src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            _feedbackRepository = feedbackRepository;
+            _kcRepository = kCRepository;
+        }
+
+        public Result<EmotionsFeedback> SaveEmotionsFeedback(EmotionsFeedback emotionsFeedback)
+        {
+            if (emotionsFeedback.Comment == null)
+                return Result.Fail("Empty Emotions Feedback");
+         
+            var knowledgeComponent = _kcRepository.GetKnowledgeComponent(emotionsFeedback.KnowledgeComponentId);
```
Comment (Luburic) [inline src/Tutor.Core/DomainModel/Feedback/FeedbackService.cs:23]: You don't need this retrieval and check for the KC. You are not performing it for the learner, and in general it is unnecessary.

**Review by Luburic — CHANGES_REQUESTED** [review]
A few minor things and the test issue and we should be done.

Btw, I've noticed the commit messages are basic. We follow semantic commit messages - https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716

**Review by Luburic — APPROVED** [review]
Looks good.

**Pull request merged**